README.md
# markdown-to-devto

A CLI tool for publish markdown articles to dev.to.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.0] - 2021-03-15
### Added
- Unit tests for python 3.6, 3.7 and 3.8.
- Ability to save file after uploading it.

### Changed
- Various small bugs, such as files not being ignored correctly.
- Support turning admonitions into quote blocks using plugin: `gatsby-remark-admonitions`.
- Support turning code blocks into proper code blocks `gatsby-remark-import-code` and `gatsby-remark-code-titles`.

## [0.2.2-beta.1] - 2020-08-19
### Changed
- Tags are converted from "react-native" to "reactnative" so it's compatible with `dev.to`.

## [0.2.1-beta.1] - 2020-05-05
### Fixed
- Don't remove lines from lists.

## [0.2.0-beta.2] - 2020-04-12
### Fixed
- Format error in README hyperlinks.

## [0.2.0-beta.1] - 2020-04-12
### Added
- Log the URL of the article that was changed/updated, fixes #1.

### Changed
- The upload and update HTTP methods now return the response JSON data.

## [0.1.0-beta.3] - 2020-03-13
### Changed
- Ignore folder logic, it checks if the ignore folder is in the path of the article.
- Upload logic into two separate functions.
- Updated the timeout time to 30 seconds.

## [0.1.0-beta.2] - 2020-03-11
### Added
- Log level to args.

### Changed
- Only upload images to Imgur we need to update article on dev.to

### Fixed
- Logger not logging to stdout.
- `query` to `params` in `http_client.py`.

## [0.1.0-beta.1] - 2020-03-10
### Added
- Initial Release.
- Publish images to Imgur and replace local references with new Imgur ones.
- Create/Update articles on dev.to.

[Unreleased]: https://gitlab.com/hmajid2301/markdown-to-devto/-/compare/release%2F0.3.0...master
[0.3.0]: https://gitlab.com/hmajid2301/markdown-to-devto/-/tags/release%2F0.3.0...release%2F0.2.2-beta.1
[0.2.2-beta.1]: https://gitlab.com/hmajid2301/markdown-to-devto/-/tags/release%2F0.2.2-beta.1...release%2F0.2.1-beta.1
[0.2.1-beta.1]: https://gitlab.com/hmajid2301/markdown-to-devto/-/tags/release%2F0.2.1-beta.1...release%2F0.2.0-beta.2
[0.2.0-beta.2]: https://gitlab.com/hmajid2301/markdown-to-devto/-/tags/release%2F0.2.0-beta.2...release%2F0.2.0-beta.1
[0.2.0-beta.1]: https://gitlab.com/hmajid2301/markdown-to-devto/-/tags/release%2F0.2.0-beta.1...release%2F0.1.0-beta.3
[0.1.0-beta.3]: https://gitlab.com/hmajid2301/markdown-to-devto/-/tags/release%2F0.1.0-beta.3...release%2F0.1.0-beta.2
[0.1.0-beta.2]: https://gitlab.com/hmajid2301/markdown-to-devto/-/tags/release%2F0.1.0-beta.2...release%2F0.1.0-beta.1
[0.1.0-beta.1]: https://gitlab.com/hmajid2301/markdown-to-devto/-/tags/release%2F0.1.0-beta.1

tests/test_cli.py
import filecmp
import time

import pytest
import requests

from markdown_to_devto.cli import check_if_we_need_to_rate_limit
from markdown_to_devto.cli import cli


@pytest.mark.parametrize(
    "args, exit_code",
    [
        ([], 2),
        (["--devto-api-key", "AKEY", "-i", "src/"], 1),
        (["--devto-api-key", "AKEY", "--imgur-id", "client-id", "-i", "src/"], 1),
        (["--devto-api-key", "AKEY", "--imgur-id", "client-id", "-m", "README.md"], 2),
        (["--devto-api-key", "AKEY", "--imgur-id", "client-id", "-f", "does_not_example"], 2),
    ],
)
def test_fail_args(runner, args, exit_code):
    result = runner.invoke(cli, args)
    assert result.exit_code == exit_code


@pytest.mark.parametrize(
    "devto_articles, args",
    [
        ([[], []], ["--devto-api-key", "AKEY", "-m", "tests/data/test.md", "-i", "src"]),
        (
            [
                [
                    {
                        "id": "123797",
                        "body_markdown": "---\nchecksum: 7eb95ce783e8896686c88c739cd30ac3\ncover_image: https://cdn-images-1.medium.com/max/1024/0*zoKBYDQXHpqWod1E.png\npublished: false\ntags: ci,react-native,gitlab\ntitle: A Test Message\n---\n\n> Before we dive straight into the CI file itself, let’s do a quicker refresher on some basic concepts. with CI/CD, Git and Gitlab CI.\n\nContinuous Integration (CI), is typically defined as making sure all code being integrated into a codebase works. It usually involves running a set of jobs referred to as a CI pipeline. Some jobs we may run include lintinusuasia tool such as Travis, Circle or even Gitlab.\n\n![a](images/a.png)\n![b](images/b.jpg)",
                        "title": "A Test Message",
                    },
                    {
                        "id": "78979",
                        "body_markdown": "---\nchecksum: dabaaec68730e589699a6006b4a73ce2\ncover_image: https://dev-to-uploads.s3.amazonaws.com/i/w00r4rpmfpjqb8wgygxu.jpg\nlicense: public-domain\ntags:\n- React Native\n- CI\n- GitLab\n- Automation\n- Android\ntitle: Auto Publish React Native App to Android Play Store using GitLab CI\n---\n\nIn this article, I will show you how can automate the publishing of your AAB/APK to the `Google Play Console`.\nWe will be using the [Gradle Play Publisher](https://github.com/Triple-T/gradle-play-publisher) (GPP) plugin to do\nautomate this process for us. Using this plugin we cannot only automate the publishing and release of our app,\nwe can also update the release notes, store listing (including photos) all from GitLab CI. \n\n**Note:** In this article I will assume that you are using Linux and React Native version >= 0.60.\n\n![c](c.jpg)\n![c](c.jpg)\n![c](c.jpg)\n\n---------------------------------------------------------------------------------------------------",
                        "title": "Auto Publish React Native App to Android Play Store using GitLab CI",
                    },
                ],
                [],
            ],
            ["-k", "AKEY", "-f", "tests/data/", "-i", "tests/data/another_folder/"],
        ),
        (
            [
                [
                    {
                        "id": "78979",
                        "body_markdown": "---\nchecksum: dabaaec68730e589699a6006b4a73ce2\ncover_image: https://dev-to-uploads.s3.amazonaws.com/i/w00r4rpmfpjqb8wgygxu.jpg\nlicense: public-domain\ntags:\n- React Native\n- CI\n- GitLab\n- Automation\n- Android\ntitle: Auto Publish React Native App to Android Play Store using GitLab CI\n---\n\nIn this article, I will show you how can automate the publishing of your AAB/APK to the `Google Play Console`.\nWe will be using the [Gradle Play Publisher](https://github.com/Triple-T/gradle-play-publisher) (GPP) plugin to do\nautomate this process for us. Using this plugin we cannot only automate the publishing and release of our app,\nwe can also update the release notes, store listing (including photos) all from GitLab CI. \n\n**Note:** In this article I will assume that you are using Linux and React Native version >= 0.60.\n\n![c](c.jpg)\n![c](c.jpg)\n![c](c.jpg)\n\n---------------------------------------------------------------------------------------------------",
                        "title": "Auto Publish React Native App to Android Play Store using GitLab CI",
                    }
                ],
                [],
            ],
            ["-k", "AKEY", "-f", "tests/data/", "-i", "tests/data/another_folder/", "-a", "my-client-id"],
        ),
        (
            [
                [
                    {
                        "id": "78979",
                        "body_markdown": "---\ncover_image: https://dev-to-uploads.s3.amazonaws.com/i/w00r4rpmfpjqb8wgygxu.jpg\nlicense: public-domain\ntags:\n- React Native\n- CI\n- GitLab\n- Automation\n- Android\ntitle: Auto Publish React Native App to Android Play Store using GitLab CI\n---\n\nIn this article, I will show you how can automate the publishing of your AAB/APK to the `Google Play Console`.\nWe will be using the [Gradle Play Publisher](https://github.com/Triple-T/gradle-play-publisher) (GPP) plugin to do\nautomate this process for us. Using this plugin we cannot only automate the publishing and release of our app,\nwe can also update the release notes, store listing (including photos) all from GitLab CI. \n\n**Note:** In this article I will assume that you are using Linux and React Native version >= 0.60.\n\n![c](c.jpg)\n![c](c.jpg)\n![c](c.jpg)\n\n---------------------------------------------------------------------------------------------------",
                        "title": "Auto Publish React Native App to Android Play Store using GitLab CI",
                    }
                ],
                [],
            ],
            ["-k", "AKEY", "-f", "tests/data/"],
        ),
    ],
)
def test_success(mocker, runner, devto_articles, args):
    result = run_cli(mocker, runner, devto_articles, args)
    assert result.exit_code == 0


@pytest.mark.parametrize(
    "args, output, expected_file",
    [
        (
            ["-k", "AKEY", "-m", "tests/data/another.md", "-o", "tests/data/tmp/"],
            "tests/data/tmp/Auto-Publish-React-Native-App-to-Android-Play-Store-using-GitLab-CI.md",
            "tests/data/expected/another.md",
        ),
    ],
)
def test_output_file(mocker, runner, args, output, expected_file):
    result = run_cli(mocker, runner, [""], args)
    assert result.exit_code == 0
    assert filecmp.cmp(output, expected_file)


def test_dev_to_api_auth_failure_get_articles(mocker, runner):
    args = ["-k", "AKEY", "-f", "tests/data/", "-i", "tests/data/another_folder/"]
    get_mock = mocker.Mock(status_code=401)
    mocker.patch("requests.get", return_value=get_mock)
    result = runner.invoke(cli, args)
    assert result.exit_code == 1


def test_os_error_image_upload(mocker, runner):
    args = ["-k", "AKEY", "-m", "tests/data/another.md", "-i", "tests/data/another_folder/", "-a", "client-id"]
    get_mock = mocker.Mock(status_code=200)
    get_mock.json.side_effect = [[], []]
    mocker.patch("requests.get", return_value=get_mock)
    mocker.patch("builtins.open", side_effect=OSError)

    articles = "---\ncover_image: https://dev-to-uploads.s3.amazonaws.com/i/w00r4rpmfpjqb8wgygxu.jpg\nlicense: public-domain\ntags:\n- React Native\n- CI\n- GitLab\n- Automation\n- Android\ntitle: Auto Publish React Native App to Android Play Store using GitLab CI\n---\n\nIn this article, I will show you how can automate the publishing of your AAB/APK to the `Google Play Console`.\nWe will be using the [Gradle Play Publisher](https://github.com/Triple-T/gradle-play-publisher) (GPP) plugin to do\nautomate this process for us. Using this plugin we cannot only automate the publishing and release of our app,\nwe can also update the release notes, store listing (including photos) all from GitLab CI. \n\n**Note:** In this article I will assume that you are using Linux and React Native version >= 0.60.\n\n![c](c.jpg)\n![c](c.jpg)\n![c](c.jpg)\n\n---------------------------------------------------------------------------------------------------"
    mocker.patch(
        "markdown_to_devto.cli.get_local_articles",
        return_value={"A Test": {"content": articles, "cover_image": "random_image.jpg", "path": "./"}},
    )
    result = runner.invoke(cli, args)
    assert result.exit_code == 0


@pytest.mark.parametrize("status_code", [401, 422, 500, 502])
def test_dev_to_api_failure_upload_articles(mocker, runner, status_code):
    args = ["-k", "AKEY", "-m", "tests/data/example.md"]
    get_mock = mocker.Mock(status_code=200)
    get_mock.json.side_effect = [[], []]
    mocker.patch("requests.get", return_value=get_mock)

    post_mock = mocker.Mock(status_code=status_code)
    mocker.patch("requests.post", return_value=post_mock)
    result = runner.invoke(cli, args)
    assert result.exit_code == 0


@pytest.mark.parametrize("exception", [requests.ConnectionError, requests.ConnectTimeout])
def test_dev_to_api_connection_failure_upload_articles(mocker, runner, exception):
    args = ["-k", "AKEY", "-m", "tests/data/example.md"]
    mocker.patch("requests.get", side_effect=exception)
    result = runner.invoke(cli, args)
    assert result.exit_code == 1


@pytest.mark.parametrize("time_wait", [34.5, 40])
def test_rate_limiting(time_wait):
    articled_uploaded, start = check_if_we_need_to_rate_limit(10, time.time() - time_wait)
    assert articled_uploaded == 0 and type(start) == float


def run_cli(mocker, runner, devto_articles, args):
    get_mock = mocker.Mock(status_code=200)
    get_mock.json.side_effect = devto_articles
    mocker.patch("requests.get", return_value=get_mock)
    create_mock = mocker.Mock(status_code=201)
    mocker.patch("requests.post", return_value=create_mock)
    create_mock.json.return_value = {"data": {"link": "https://imgur.com/123456"}, "url": "random_url.com"}
    mocker.patch("requests.put", return_value=create_mock)
    result = runner.invoke(cli, args)
    return result


src/markdown_to_devto/http_client.py
# -*- coding: utf-8 -*-
r"""A client used to send API requests to dev.to. It can be used to get current articles. It can also be used to
update and create new articles.

Example:
    ::

        $ devto = DevtoClient(api_key=12345678)

.. _Google Python Style Guide:
    http://google.github.io/styleguide/pyguide.html

"""
import os

import requests

from .utils import exceptions


class HTTPClient:
    def __init__(self, devto_api_key=None, imgur_client_id=None):
        self.devto_api_key = devto_api_key
        self.imgur_client_id = imgur_client_id

    def get_articles(self):
        """Gets all the articles published on dev.to under your account.
        The API uses pagination and only returns 30 articles at a time. So
        we keep querying the API and incrementing the page number until the number
        of articles is less than 30.

        Returns:
            dict: Where the key is the article title and values are the info about the article.

        """
        page = 1
        articles_data = {}

        url = "https://dev.to/api/articles/me/all"
        while True:
            current_articles = self._make_http_request(
                method="get", url=url, params={"page": page, "per_page": 200}, headers={"api-key": self.devto_api_key}
            )
            if not current_articles:
                break

            page += 1
            for article in current_articles:
                title, uid, content = (article["title"], article["id"], article["body_markdown"])
                articles_data[title] = {"id": uid, "content": content}

        return articles_data

    def update_article(self, article_id, article_data):
        """Update an already existing article on dev.to.

        Args:
            article_id (int): The article id on dev.to we are trying to update.
            article_data (dict): The "new" article data we are updating on dev.to, such as content.

        """
        data = {"article": {"body_markdown": article_data["content"]}}
        url = f"https://dev.to/api/articles/{article_id}"
        headers = {"api-key": self.devto_api_key}
        response = self._make_http_request(method="put", url=url, json=data, headers=headers)
        return response

    def create_article(self, article_data):
        """Create a new article on dev.to.

        Args:
            article_data (dict): The article data we are creating on dev.to, such as content.

        """
        data = {"article": {"body_markdown": article_data["content"]}}
        url = "https://dev.to/api/articles"
        headers = {"api-key": self.devto_api_key}
        response = self._make_http_request(method="post", url=url, json=data, headers=headers)
        return response

    def upload_image(self, local_path):
        """Uploads an image to Imgur (as an anonymouse image).

        Args:
            local_path (str): The path to the image you want to upload..

        Returns:
            str: URL on imgur where the image was uploaded.

        """
        url = "https://api.imgur.com/3/upload"
        headers = {"Authorization": f"Client-ID {self.imgur_client_id}"}
        image_path = os.path.join(os.getcwd(), local_path)
        response = self._make_http_request(
            method="post", url=url, headers=headers, files={"image": open(image_path, "rb")}
        )
        link = response["data"]["link"]
        return link

    def _make_http_request(self, method, url, **kwargs):
        """Make a HTTP request to dev.to, for example to.

            - To get articles
            - To update article
            - To create an article

        Args:
            method (str): The HTTP method/verb to use i.e. "post", "get".
            url (str): The URL/endpoint to send the HTTP request to.
            **kwargs: Extra parameters to use with "requests" such as `query` or `json`.

        Raises:
            HTTPConnextionException: When there are connection issues or the request times out.

        """
        http_method = getattr(requests, method)
        try:
            response = http_method(url, timeout=30, **kwargs)
        except (requests.ConnectTimeout, requests.ConnectionError) as e:
            raise exceptions.HTTPConnectionException(msg=e)

        data = response.json()
        self._handle_response(status_code=response.status_code, response_json=data)
        return data

    @staticmethod
    def _handle_response(status_code, response_json):
        """Checks the HTTP response from the dev.to API and raises exceptions if the status code is
        not what we expect.

        Args:
            status_code (int): The HTTP response status code.
            response_json (dict): The response data sent from the dev.to API.

        Raises:
            HTTPBadRequestException: If the API returns a "bad" HTTP response (400, 422).
            HTTPAuthRequestException: If the API returns an auth error HTTP response (401).
            HTTPServerException: If the API returns an server error HTTP response (5xx).

        """
        if status_code in [400, 422]:
            raise exceptions.HTTPBadRequestException(msg=response_json)
        elif status_code == 401:
            raise exceptions.HTTPAuthException(msg=response_json)
        elif status_code not in [200, 201]:
            raise exceptions.HTTPServerException(msg=response_json)
